hail/hail/src/is/hail/expr/ir/IR.scala

```diff
@@ -26,25 +25,21 @@ import java.io.OutputStream
 import org.json4s.{DefaultFormats, Extraction, Formats, JValue, ShortTypeHints}
 import org.json4s.JsonAST.{JNothing, JString}
 
-sealed trait IR extends BaseIR {
+trait IR extends BaseIR {
   private var _typ: Type = null
 
-  def typ: Type = {
-    if (_typ == null)
+  override def typ: Type = {
+    if (_typ == null) {
       try
         _typ = InferType(this)
       catch {
         case e: Throwable => throw new RuntimeException(s"typ: inference failure:", e)
       }
+      assert(_typ != null)
```

Comment:
Is `null` a valid return from `InferType`?

Reply:
No, `InferType` should assign a type to every node. I believe I added this because I ran into a bug where a type did remain null. I don't remember what happened, but since `InferType` often just lifts types from children, one potential failure mode is to call `InferType` on a node when some of its children still have null types.